Project proposal for Conformance Validation

mayankshah1607 opened the pull request.

mayankshah1607 committed:
Initial commit

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Summary, Problem Statement and Unresolved Questions

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Summary

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Step 2

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Follow-up for Step 2

- Add points to support why a certain approach might be taken to answer "why?" questions

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update label selector PR info

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+# Summary
+
+[summary]: #summary
+This project proposes a new test suite that shall be used for conformance validation. The new test suite shall be used to validate non-trivial network communication (HTTP, gRPC, websocket) among stateless and stateful workloads in the Linkerd data plane. The correctness of the interaction between the Linkerd control plane and the Kubernetes API Server will also be tested. This shall be done by ca ...
```
s/linkerd/Linkerd

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+## Interaction with other features
+The e2e conformance tests proposed would be carried out outside the Linkerd CI. These tests would make use of the kubectl and linkerd binaries to interact with the cluster and with various features Linkerd can perform. The testing process can be configured easily using a .yaml definition file (more details explained under “implementation details”). Such an approach gives the users the flexibility ...
+
+This project also proposes an optional sample distributed application - “MovieChat” that can exercise various features of Linkerd and also surface issues regarding various APIs, protocols and topologies such as streaming, websockets, MySQL and Redis. Currently, Linkerd does not have a sample application that implements these and would benefit from testing for conformance. We initially start off b ...
```
If you are planning to start your test suite implementation with emojivoto, then I recommend moving the reference to the MovieChat app to the "Bonus" (or "nice to have") section.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+This project also proposes an optional sample distributed application - “MovieChat” that can exercise various features of Linkerd and also surface issues regarding various APIs, protocols and topologies such as streaming, websockets, MySQL and Redis. Currently, Linkerd does not have a sample application that implements these and would benefit from testing for conformance. We initially start off b ...
+
+## Implementation Details
+The test suite shall be executed as a job on the cluster using [Sonobuoy](https://sonobuoy.io/). It is a great tool when it comes to conformance validation, and its [plugin model](https://sonobuoy.io/docs/master/plugins/) can allow us to easily extend functionality beyond k8s conformance validtion. Sonobuoy also saves developers the overhead of configuring the desired objects such as Namespaces,  ...
```
I prefer the default mode to not have a hard dependency on Sonobuoy. Those content (as seen below) should appear first.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+[design-proposal]: #design-proposal
+
+## Interaction with other features
+The e2e conformance tests proposed would be carried out outside the Linkerd CI. These tests would make use of the kubectl and linkerd binaries to interact with the cluster and with various features Linkerd can perform. The testing process can be configured easily using a .yaml definition file (more details explained under “implementation details”). Such an approach gives the users the flexibility ...
```
When referring to Linkerd as a project, please use uppercase. When referring to the CLI, please backtick "linkerd".

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+In the Dockerfile, we define that the following must happen in the container running as a job:
+- Setup and install all the binaries/tools required by the testing environment - linkerd, kubectl, mysql, etc.
+- Copy the test configuration file conformance_config.yaml which is responsible for describing how the tests must run (see below for implementation details).
+- Copy _run.sh_ file that initiates the tests using the go test cmd.
```
Backtick "go test cmd", since it's a command.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+- Copy the test suite
+- ENTRYPOINT / CMD that executes _./run.sh_
+
+> Note: When generating a sonobuoy plugin, the plugin definition file sets the start-up command as ./run.sh by default. If we set the container’s ENTRYPOINT/CMD as go test, this may get overridden to ./run.sh because of how sonobuoy generates plugins. The generated plugin definition file may however be modified to not override the startup command of the container, but this may have a negative imp ...
```
s/sonobuoy/Sonobuoy
Backtick "go test"

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+**Bonus**
+
+If the users do not want to do any of this, we can provide a default docker image with some predefined test configurations - Eg : _gcr.io/linkerd.io/conformance:v1.0.0_, and also host a plugin definition file that makes use of this image - _https://run.linkerd.io/path/to/plugin.yaml_. This approach is great if users quickly want to get the conformance validation tool up and running.
```
s/docker/Docker

I am not  sure what this plugin definition file (`plugin.yaml`) is. How is it related to `conformance_config.yaml`?

mayankshah1607 replied:
The `plugin.yaml` is the file that defines how and which plugin Sonobuoy will run. This file is required with the `sonobuoy run` cmd. It is not related to `conformance_config.yaml` - this file is only required by our testing tool. 
I've added more details about this in the RFC. Please let me know if this resolves it. :smile:

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,449 @@
+$ sonobuoy run --plugin https://run.linkerd.io/path/to/plugin.yaml
+```
+
+Users shall also be allowed to run these tests without having to depend on Sonobuoy. Much like the [Kubernetes conformance validation process](https://github.com/kubernetes/kubernetes/tree/master/cluster/images/conformance) we can provide a yaml configuration file that can be used to run a pod that runs these tests on the cluster. This yaml configuration file is responsible for setting up things  ...
````
s/yaml/YAML

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+### Test Configuration
+The conformance test suite proposed in this project shall be made customizable as per the users’ preferences, which they may mention in a .yaml file. Such an approach not only enhances user experience but can also give the testers greater insights on their cluster and Linkerd.
+
+This configuration file shall be unmarshalled into an object during runtime so that tests may read the desired properties from it and can run accordingly. This feature gives users the flexibility of running tests according to their preference. Proposed structure : 
```
s/preference/preferences

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+Users shall also be allowed to run these tests without having to depend on Sonobuoy. Much like the [Kubernetes conformance validation process](https://github.com/kubernetes/kubernetes/tree/master/cluster/images/conformance) we can provide a yaml configuration file that can be used to run a pod that runs these tests on the cluster. This yaml configuration file is responsible for setting up things  ...
+
+### Test Configuration
+The conformance test suite proposed in this project shall be made customizable as per the users’ preferences, which they may mention in a .yaml file. Such an approach not only enhances user experience but can also give the testers greater insights on their cluster and Linkerd.
```
Is this config YAML file the same as the `conformance_config.yaml` described above, which `run.sh` relies on? It feels like a lot of work to read and unmarhsal this YAML in the `run.sh` shell script. Can you make `run.sh` accept command line arguments?

mayankshah1607 replied:
`run.sh` does not depend on `conformance_config.yaml`. The latter YAML file is required by our testing code - it is unmarshalled in go, not in shell script. This approach is very similar to https://github.com/linkerd/linkerd2/issues/3870
I have added a note on how we can make `run.sh` accept command line arguments and parse them to our go code. However, considering the increasing number of flags we might end up having, a YAML definition file seems like a better option. Please have a look at the newly added note and let me know what you think :smile:

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+**1. Auomatic Proxy Injection**
+
+Proxy injection process works by adding a `linkerd.io/inject: enabled` annotation to pod template / namespace. This triggers an admission webhook that injects the `linkerd-proxy` container to the targeted deployments. This process shall be tested in 7 steps:
+- Check if `inject` command (with and without `--manual` flag) outputs the desired configuration with the appended annotation.
```
Are you planning to validate the entire YAML blob output of the `inject` command? Or just specific fields? What are you comparing the output with, to verify if it's correct?

mayankshah1607 replied:
In case of emojivoto, comparing against a predefined YAML seems fine. But if we're moving outside of emojivoto, we have to use the `client-go` library. Added a note on this, please let me know if that resolves it. :smile:

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+- Check if `inject` command (with and without `--manual` flag) outputs the desired configuration with the appended annotation.
+- Wait for the pods to come to a _Running_ state for a certain duration, after which the test shall fail.
+- Check if the pods actually have the `linkerd-proxy` and the `linkerd-init` container
+- Check if the linkerd-proxy container is reachable by sending a `GET` request to liveness/readiness probe - to ensure service discovery is functional.
```
Backtick "linkerd-proxy" to be consistent with the previous bullet point.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+- Check if application deployment has desired replicas
+- Get application pods IPs
+- Issue linkerd edge cmd
+- Validate if the desired edges are returned
```
`linkerd edges` tells you whether the link between two workloads is mTLS'ed. If not, it provides the reasons. Check out `linkerd -n <ns> edges <kind> -o json`.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+- Validate the returned output by counting instances of desired route substrings using `strings.Count`
+
+**6. Distributed Tracing**
+Much like the emojivoto application, our MovieChat application shall be configured with the OpenConsensus Agent library to support tracing information in the requests. The following checks must be carried out:
```
s/OpenConsensus/OpenCensus

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+**6. Distributed Tracing**
+Much like the emojivoto application, our MovieChat application shall be configured with the OpenConsensus Agent library to support tracing information in the requests. The following checks must be carried out:
+- Install the collector and check if it is up and running in the tracing namespace.
+- Install jaeger and check if it is up and running in the tracing namespace.
```
s/jaeger/Jaeger

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+It is essential to test ingress to ensure that the ingress controller re-writes incoming headers to the internal service name. This process ensures that service discovery is working. This test shall include deploying various types of ingress controllers (Nginx, Traefik and Ambassador) :
+- Check and install flagger and wait until a certain duration for the status of the deployment to change to “Running”, after which the test fails.
+- Deploy an ingress control. An example from the docs:
```
To confirm, the YAML below only deploys an ingress resource. You will need to include the YAML of the ingress controller in your test tool

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,449 @@
+
+The movie chat application or emojivoto shall leverage gRPC streaming. It is essential to ensure that streaming services are not affected by Linkerd - under normal conditions as well as during stress tests. To test gRPC streaming, we shall make use of a curl replacement for gRPC called grpcurl (https://github.com/fullstorydev/grpcurl). grpcurl is a command-line utility like curl that can easily i ...
+```bash
+grpcurl -plaintext <url>:<port> ServiceName.MethodName
````
You probably don't need `grpcurl` and `curl`. If there is constant traffic between the pods, `linkerd stat` will show the success rate, which should be 100%.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+**3. MySQL and Redis***
+
+Port-forwarding can be initiated on the pods serving these applications. Once done, the tests shall make sample queries to these to ensure that appropriate output is returned. Additionally, users may be allowed to mention if these pods must be injected or not in the test configuration file.
```
Why not just modify emojivoto to have persistence, to avoid needing a port-forward to send queries to the DB? The vote-bot will generate continuous traffic to the backend, which can be set up to simulate cache hit (Redis) and a cache miss (MySQL).

mayankshah1607 replied:
Done. Added more info on this.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+## Corner Cases
+
+It is possible to have prior knowledge regarding which use-cases may need to be configured differently depending on the environment. For eg, tap requires extra RBAC configurations on GKE. Instead of having the corresponding tests for tap failing, users may be allowed to explicitly mention the environment these tests shall run on, so that test cases may take a different path, i.e the tests may be  ...
```
s/use-cases/use cases
s/For eg/E.g.

> so that test cases may take a different path,

Or just document it so that GKE users are aware of preconditions that must be satisfied.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+6. gRPC streaming (client/server/bi-directional streaming and unary)
+7. Websockets
+8. MySQL
+9. Redis
```
You already talked about this above; feels redundant.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+## Goals
+The goal of this project is to :
+1. develop a sample application (that comprises microservices) that can exercise various features of Linkerd, mainly that involve the data plane. (from https://github.com/linkerd/linkerd2/issues/2079) This application shall also exercise various protocols and topologies that could benefit from deeper testing and surface issues around streaming, websockets, MySQL and Redis.
+2. provide an extensive e2e test suite that can make use of this application and validate many properties of the cluster, to help a user test for conformance w.r.t Linkerd, i.e to check if the cluster has been configured to work correctly over a longer period of time.
```
Feels like repeated content.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+    6. gRPC streaming (client/server/bi-directional streaming and unary)
+    7. Websockets
+    8. MySQL
+    9. Redis
```
There are some items in here that I feel it's less important (either because they are covered by the integration test or they don't tell us much), and those under "Good-to-have" which I think are crucial. Here's my wish list of test cases:

1. Proxy injection - this should check for deployment errors due to unforeseen configuration like PodSecurityPolicy, LimitRanger, RBAC, service account token automount is disabled etc.
1. Linkerd `tap`, `stat` (latency), `routes` (service profile), `edges` (mTLS) commands
1. `tap` extension API server is functional (`linkerd check --pre` can do that)
1. Retries and timeouts (take a look at `test/serviceprofiles/service_profiles.go` to see if you can reuse code)
1. Ingress traffic from ingress controller to the pod
1. Data plane proxies are healthy (start with `linkerd -n <ns> check --proxy`, then drill down into 503 errors in logs)

Other less important items:

1. MySQL, Redis where you end up having to skip some ports anyway. It will still be important to know that these databases still work once their ports are skipped
1. gRPC streaming, websocket etc. shouldn't make a difference

There are already some integration tests for tracing. Since canary relies on a number of external components, I'm concerned that the tests might be flaky. It feels more efficient to perform canary testing manually, than to write code with assumptions on external components' behaviour.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+As a Linkerd user, it is important to check if Linkerd works as expected even outside of the emojivoto / any other sample application. Setting the workloadNamespace property in the test configuration file gives the users this flexibility :
+- During runtime, the value for this property is picked up from the unmarshalled object, and a `config.linkerd.io/e2e: conformance` label is appended to the namespace config
+- The tests then hit the k8s API with a label selector to annotate all objects (deployments) under this namespace with the `linkerd.io/inject: enable` annotation to trigger auto-injection.
```
s/enable/enabled

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+As a Linkerd user, it is important to check if Linkerd works as expected even outside of the emojivoto / any other sample application. Setting the workloadNamespace property in the test configuration file gives the users this flexibility :
+- During runtime, the value for this property is picked up from the unmarshalled object, and a `config.linkerd.io/e2e: conformance` label is appended to the namespace config
+- The tests then hit the k8s API with a label selector to annotate all objects (deployments) under this namespace with the `linkerd.io/inject: enable` annotation to trigger auto-injection.
+- It is crucial for linkerd cmds like stat, route and edge to support label selectors, hence, a patch for this must be made.
```
s/linkerd/Linkerd
s/cmds/commands
Backtick all the commands

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,449 @@
+
+[future-possibilities]: #future-possibilities
+
+As a Linkerd user, it is important to check if Linkerd works as expected even outside of the emojivoto / any other sample application. Setting the workloadNamespace property in the test configuration file gives the users this flexibility :
```
Try with "As a Linkerd user, it is important to know that Linkerd works with my services, not just emojivoto".

Backtick "workloadNamespace"

ihcsim reviewed (commented):
@mayankshah1607 See comments below. In particular, the _Deliverables_ section around the prioritization of the test cases.

mayankshah1607 committed:
Address comments on review by @ihcsim

- s/linkerd/Linkerd where
- s/cmds/commands
- s/sonobuoy/Sonobuoy
- s/yaml/YAML
- s/docker/Docker
- s/preference/preferences
- s/jaeger/Jaeger
- Move "MovieChat" summary to "Nice to have"
- Add more info regarding the default Sonobuoy plugin file and how they are related to
  conformance_config.yaml (Under "Nice to have")
- Add backticks where required
- Add note on how test config file depends on go code and mention responsibily of run.sh
- Add note on using command line arguements for run.sh
- Add details on how proxy injections are validated
- Update `linkerd edges` validation using `-o json` flag
- Add information regarding ingress controllers and its YAML
- Update gRPC testing method
- Update MySQL and Redis testing method
- Update "Corner cases" section
- Shorten "Goals" section
- Shorten "Use cases" section
- Update "Deliverables"
- Update "Future possibilities"
- Add info on data plane health checks
- Add description for retries and timeouts

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update test configuration

- Add information of how the absence of test configuration file shall be handled

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fix typos

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,484 @@
+- From the `linkerd-proxy` container of each of the pods, check for 503 errors.
+
+**9. `tap` extension API server**
+- Issue a `check` command - `linkerd -n <ns> check --pre -o json`
```
I've left this bit incomplete as I wasn't sure which check can satisfy this test. Could you give me a little more info on this?

ihcsim replied:
Some custom Kubernetes clusters don't always have the [aggregation layer configured](https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/#enable-kubernetes-apiserver-flags), causing the `tap` service which is an extension API server to fail. `linkerd check --pre` can now detect this problem. See https://github.com/linkerd/linkerd2/pull/4062

mayankshah1607 reviewed (commented):
@ihcsim Thanks for your valuable feedback! I've addressed all the comments, and left a reply on some - could you have a look and let me know if the new changes resolves those comments?
I wanted to know if I should still keep details about Distributed tracing and Canary release under "Testing methodologies" section as they're no longer that important?
Also had a question regarding `tap` extension API server which you can find in the comment below.
Thanks! :smile:

Pothulapati commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,484 @@
+```bash
+# setup docker registry
+$ export USER=<your public registry>
+$ docker build . -t $USER/progress-reporter:v0.1
````
Hmm, Asking users to build the image just to pass some configuration, feels a bit hard to me. Can we use something like configmap attaching as volume, etc to make the container read the config during runtime? Yes I see that the plugin yaml is being generated later in the pipeline. Can we use kustomize to override the generated yaml with attaching a configmap?

If we don't find anything to replace, we can go with this way I guess!

mayankshah1607 replied:
As per @ihcsim 's suggestion, I have removed this part and instead made the `.run.sh` file accept a list of arguments. Could you take a look? :smile:

ihcsim commented:
> I wanted to know if I should still keep details about Distributed tracing and Canary release under "Testing methodologies" section as they're no longer that important?

You can summarize them in one sentence. If we decided to do them over the summer, we can open a new PR to amend this RFC.

Also, two more suggestions:

The "Implementation Details" section seems to go back-and-forth between non-Sonobuoy and Sonobuoy modes, which is confusing. Please group all the information together under subsections. E.g., start the section with the "default mode" to run the test (I don't want Sonobuoy to be a hard dependency). Then follow by a subsection with just information and steps on running the tests as Sonobuoy plugin.

Also, consider removing the subsection on the custom MovieChat application. It makes your RFC longer, without adding much value, since we are starting with an existing application like emojivoto. If we decided to build that custom application over the summer, we can open a PR against your RFC to add it back in.

mayankshah1607 committed:
Remove MovieChat section

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Distributed tracing and Canary release

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor test methodologies section

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,484 @@
+We initially start off by modifying the existing emojivoto application to have feature flags to enable/disable features such as MySQL, Redis, gRPC, websockets, etc. This is important because emojivoto is heavily used in the getting started process.
+
+## Implementation Details
+The test suite and its dependencies shall be packaged as a Docker container. This container shall run as a job in the cluster. In the Dockerfile, we define that the following must happen :
```
Outside of Sonobuoy, is it necessary to run your test framework as a container?

ihcsim replied:
s/happen :/happen:

mayankshah1607 replied:
> Outside of Sonobuoy, is it necessary to run your test framework as a container?

Nope. I've revamped this section and clearly described how the test framework will work with and without Sonobuoy.

Pothulapati replied:
I don't have strong opinions here, but I always liked having a `docker run <something>` when its a one-time thing. Cloning the repo, and then running the tests on my local env ain't that hard, but having a container where they can pass both `kubeConfig` and configuration, and letting it run the tests feels really awesome without effecting my local env. WDYT?

mayankshah1607 replied:
@Pothulapati That's actually possible with the current approach. The Sonobuoy section talks about having a Dockerfile (which is a requirement for Sonobuoy). This Dockerfile can definitely be utilised to run the tests locally as a part of the default mode - simply by building the image and then running it by passing the `./run.sh` args (as `ENTRYPOINT` is set to `./run.sh`) as you mentioned.

I wasn't sure how much information should have been added because I didn't want this section to extend for too long. Nevertheless, I have now added info about under the section that introduces the Dockerfile. Do have a look and let me know if that's what you're looking for! :smile: 
The RFC also talks about hosting this docker image on a public repository, users may directly pull from there as well!

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,484 @@
+```bash
+# setup docker registry
+$ export USER=<your public registry>
+$ docker build . -t $USER/progress-reporter:v0.1
````
What is `progress-reporter`? If you are planning to reuse this [one](https://github.com/vmware-tanzu/sonobuoy/tree/master/examples/plugins/progress-reporter), please link to it, and explain what it does.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,484 @@
+   # RBAC configuration to be used if on GKE
+   # If left empty, use default RBAC settings as
+   # mentioned here
+   gkeRbacConfig: <string>
```
The only `testOptions` I find useful from the list are `wait` and `enableTest`. What about others like `failFast`, `buildTags`, `parallel` etc.?

As for `installOptions`, I like to do as little work as possible to get the arguments of my existing `linkerd install args....` command into your test framework. These are some of the commands that I use often:
```
linkerd install | kubectl apply -f -

linkerd install [--ha] --ignore-cluster | kubectl apply -f -

linkerd install --linkerd-cni-enabled | kubectl apply -f -

linkerd install --identity-trust-anchors-file=ca.crt --identity-issuer-certificate-file=issuer.crt --identity-issuer-key-file=issuer.key --identity-issuance-lifetime=1m  | kubectl apply -f -
```

In my ideal world, I like to just use `ctrl+r` to find a previous command, copy its arguments, and append and paste it into `./run.sh args......`. 

I am not keen on the idea of having to keep a config file around, where it requires me to translate `--identity-trust-anchors-file`, `--identity-issuer-certificate-file` etc. into whatever keys/variables that file can support. How do I know what `--identity-trust-anchors-file` is in your YAML config? If there is a typo, will your `run.sh` be able to help me locate it?

Personally, I am perfectly ok with something like:
```
./run.sh --wait=30s \
    --enableTest="TestAutoProxyInjection/*" \
    --failFast \
    --identity-trust-anchors-file=ca.crt \
    --identity-issuer-certificate-file=issuer.crt \
    --identity-issuer-key-file=issuer.key \
    --identity-issuance-lifetime=1m \
    --proxy-cpu-limit=1 \
    --proxy-cpu-request=100m \
    --proxy-memory-limit=250Mi \
    --proxy-memory-request=20Mi \
    --ha \
    --addon-config=addon-config.yaml
```
It's likely that I will end up adding `run.sh` to some existing test scripts I have anyway.

mayankshah1607 replied:
That makes sense. I've removed the test configuration section and instead added details on how we'll go about using these flags to build a `go test` command. 
Could you have a look and let me know if this looks good? :smile:

mayankshah1607 committed:
Revamp implementation details

- Group information regarding sonobuoy approach and vanilla approach
- Remove test configuration
- Introduce command line flags for run.sh

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Minor edits

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update deliverables

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update dependencies

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Unresolved Question

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add `workloadNamespace` flag

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Minor edits

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add tap API server check

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update run.sh flags

- Add new `--skipInstall` flag
- Add info on command builder in script

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Minor edits to testing methodology and links

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,443 @@
+## Interaction with other features
+The e2e conformance tests proposed would be carried out outside the Linkerd CI. These tests would make use of the `kubectl` and `linkerd` binaries to interact with the cluster and with various features Linkerd can perform.
+
+We initially start off by modifying the existing emojivoto application to have feature flags to enable/disable features such as MySQL, Redis, gRPC, websockets, etc. This is important because emojivoto is heavily used in the getting started process.
```
A Note to remember: Care has to be taken to make these features that we will add are independent of each other, and would not fail the application, when any of the flag is turned off i.e Ideally each new protocol/feature should be independent of the normal working of the application

Pothulapati commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,443 @@
+1. A `run.sh` file that accepts a list of flags (see use case and implementation details below) and calls `go test` accordingly.
+2. The test suite (the `_test.go` files)
+
+To run the test suite, users will be required to execute `./run.sh` with a bunch of flags (more details below). Further, the output of the tests shall be written to a filepath passed as a flag to `./run.sh`.
```
we can have a `-o` flag, and the default can be `stdout` for the output.

Pothulapati commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,443 @@
+  - --enableTest="TestAutoProxyInjection/*" \
+  - --failFast \
+  - --ha \
+  - --addon-config=addon-config.yaml
```
When doing the conformance in this way, How are users going to overwrite `addon-config`, as it would be looking inside the running job container for the file, right?

mayankshah1607 replied:
Yes, this would actually only be an issue while running on Sonobuoy. To tackle that, I've added a new `./run.sh` flag - `--addon-config-sonobuoy`.
The RFC talks about running the tests in 2 ways:
1. Default mode (user's local environment or a container) - here, we may provide a default `addon.yaml` that the users can edit, and because its on a local environment, `linkerd install` can directly accept it. 
In case of running it in a container, the users include the `addon.yaml` file in the build process of the Dockerfile we provide.

2. Sonobuoy - We use the previously mentioned `--addon-config-sonobuoy` flag. Under the hood, it takes your local `addon.yaml` file and copies it onto the Sonobuoy pod using `kubectl cp`

ihcsim replied:
Just put the `addon.yaml` in a folder (e.g., `config`, `etc`, `data`), and update your `Dockerfile` to always `COPY` everything in that folder.

Pothulapati commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,443 @@
+```yaml
+sonobuoy-config:
+  driver: Job
+  plugin-name: l5d-conformacne
````
s/conformacne/conformance

mayankshah1607 committed:
Fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add addon flag info

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update info on Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update formatting

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+We initially start off by modifying the existing emojivoto application to have feature flags to enable/disable features such as MySQL, Redis, gRPC, websockets, etc. This is important because emojivoto is heavily used in the getting started process.
+
+## Implementation Details
+The proposed test framework shall be written using Go. Apart from managing outputs, the standard `testing` framework provides us features that our conformance test suite could leverage, for e.g - running tests in parallel, passing command line flags to make tests configurable, skipping tests, etc. Moreover, this shall also give us the flexibility to reuse existing code (packages, objects, etc.) f ...
```
>  from the _linkerd2_ codebase

Hyperlink linkerd2 to point to the GH repo.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+1. A `run.sh` file that accepts a list of flags (see use case and implementation details below) and calls `go test` accordingly.
+2. The test suite (the `_test.go` files)
+
+To run the test suite, users will be required to execute `./run.sh` with a bunch of flags (more details below). Further, the output of the tests shall be written to a filepath passed as a flag to `./run.sh`. By default, `stdout` shall be used to show output of the tests.
```
s/bunch/group

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+--addon-config String
+    Filepath to the addon config file. This flag must be used only in "default mode"
+
+--addon-config-sonobuoy String
```
I don't think this is necessary. See comment above on adding a `COPY` to the Dockerfile.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
````diff
@@ -0,0 +1,477 @@
+--workloadNamespace String
+    This could be a future possibility. The value of this flag identifies the namespace of the workloads against which the conformance tests must run.
+```
+Internally, the script builds a `go test` command by simply appending these values to it. The test configuration specific flags (such as `--wait`,`--enableTests`,`--workloadNamespace`, etc) are passed down to the test suite using the _flags_ go module. These values may be stored in variables (instance of objects, for e.g `TestOptions`) so that tests may run accordingly. A universal `TestHelper` o ...
````
s/builds/executes

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+-  `--ha`
+-  `--addon-config`
+
+The script also builds a `linkerd install` command by simply appending these flags to it. The built command is then executed to install linkerd on the cluster.
```
s/builds/executes
s/linkerd/Linkerd

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+4. Check if Linkerd control plane is installed
+5. Install the sample application
+
+Once the test suite is complete (success / failure ) all resources created for the test are destroyed if not immediately after each test.
```
I prefer to keep the resources around if the tests failed so I can debug them.

mayankshah1607 replied:
Hmm. But what if the tests are successful? I think it would be nice to automatically clean up the resources in that case. However, is it possible that sometimes users may just want to clean up anyway regardless of the failure of success of the tests? We could then have a `--clean` flag that indicates that the resources must always be cleaned.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+
+**linkerd stat**
+
+- Check if prometheus pod(s) is/are available
```
s/prometheus/Prometheus

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+- Issue `linkerd routes` cmd
+- Validate the returned output by counting instances of desired route substrings using `strings.Count`
+  
+**3. Ingress**
```
This is a good test to have. Two things to keep in mind:

1. If this test fails, the test needs to be able to report useful information back to the users, to aid with debugging. Or maybe we can just keep the cluster around so users can e.g. retrieve the ingress controller logs (in debug mode)
1. Local K8s clusters won't return an external IP

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+- Make a `GET` request to the `linkerd-proxy` containers (of each of the pods) at the `/metrics` (Liveness probe) and `/ready` (Readiness probes) to ensure that they are reachable.
+- From the `linkerd-proxy` container of each of the pods, check for 503 errors.
+
+**5. `tap` extension API server**
```
This test suite should be run sooner. Otherwise, your "essential linkerd command" test suite won't pass.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+- Issue a `check` command - `linkerd -n <ns> check --pre -o json`
+- Validate the returned JSON by ensuring that the array the check with description `"can read extension-apiserver-authentication configmap"` under category `"pre-kubernetes-setup"` says `"success"`
+
+**6. Retries and Timeouts**
```
Something to think about is that instead of relying on a property like `effective_success` to determine the test assertion, which has proven to be flakey, it will be nice if there is a way to record/observe the occurrence of retries on the server side, and timeouts on the client side.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+
+## Use Cases
+
+This standalone testing tool shall be used for verifying the features of Linkerd involving the data plane, as listed under the "Testing Methodologies for various features" section.
```
This section is repetitive.

ihcsim commented on rfc/2020/conformance-validation/mayankshah1607/rfc.md:
```diff
@@ -0,0 +1,477 @@
+    4. Distributed tracing
+    5. MySQL, Redis
+   
+- A new sample application - MovieChat - that has all the features required from conformance validation perspective
```
I recommend that you move these 3 sections to the top of the RFC. Maybe add a _Goals and Deliverables_ parent section between the Phase 1 and Phase 2 sections. This provides your readers with a clear picture of what you will be delivering, before going into the implementation details in the Phase 2 section.

mayankshah1607 committed:
Address @ihcsim's review

- Move Goals, Non-goals and Deliverables section on top of the RFC
- Add linkerd2 hyperlink
- Add new `--clean` flag
- Remove `--addon-config-sonobuoy` flag
- Bring tap API test on top
- s/prometheus/Prometheus

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Goals section

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add info on retries/timeouts monitoring

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update formatting and typos

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add details on monitoring occurences of retries/timeouts

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>
